### PR TITLE
Basic Sharepoint online connector

### PR DESCRIPTION
**Pull request by artem-shelkovnikov**
## Part of https://github.com/elastic/enterprise-search-team/issues/4701

This PR adds a working Sharepoint Online connector.

Most of features work - such as extraction of content, simple advanced filtering rules, proper validation/error messaging, handling of throttling and such.

Connector has decent code coverage that can be improved with subsequent PRs.

`ftest` works but not perfectly - only generates static data of really small amount. This will be addressed with subsequent PRs.

This PR should be merged ASAP because other teams depend on this connector code being in - without it it'll be much harder to merge everything later.

I will be improving connector in the meantime anyway - addressing left TODOs, improving test coverage, providing an ftest that serves a lot of dynamic documents and which can be used for ftests.

Please block merging only if _critical_ problems that break `main` are found, otherwise please report problems and I'll address them with further work.

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

## Release Note

Added Sharepoint Online connector.

### CODE REVIEW

File: connectors/sources/sharepoint_online.py [hunk 1]
```diff
@@ -0,0 +1,589 @@
...
+
+                        download_func = None
+
+                        yield list_item, download_func
+
+                async for site_page in self._client.site_pages(site["webUrl"]):
+                    site_page["_id"] = site_page["GUID"]
+                    site_page["object_type"] = "site_page"
+                    if "LayoutWebpartsContent" in site_page and site_page["LayoutWebpartsContent"]:
+                        site_page["LayoutWebpartsContent"] = BeautifulSoup(site_page["LayoutWebpartsContent"], features="html.parser").get_text()
```
Comment (tarekziade) [inline connectors/sources/sharepoint_online.py:522]: That is going to be cpu-bound and blocking the GIL. Can you put the parsing in a separate function in utils. I'll explore alternatives to see of we can avoid the toll

Comment (tarekziade) [reply connectors/sources/sharepoint_online.py:522]: Added a note there -- if this data is an xml-compatible tree you can switch to `lxml` https://github.com/elastic/connectors-python/issues/973

Comment (artem-shelkovnikov) [reply connectors/sources/sharepoint_online.py:522]: Sure, this code is there in WIP state and not indicative of final solution :D

I'll move it to a function in utils and explore the suggestion in #973.

File: connectors/sources/sharepoint_online.py [hunk 2]
```diff
@@ -0,0 +1,740 @@
...
+        )  # TODO: make configurable, that's a scary property
+
+        # Graph API and Sharepoint API scroll over slightly different fields:
+        # - odata.nextPage for Sharepoint REST API uses
+        # - @odata.nextPage for Graph API uses - notice the @ glyph
+        # Therefore for flexibility I made it a field passed in the initializer,
+        # but this abstraction can be better.
+        self._scroll_field = scroll_field
+
+    async def fetch(self, url, debug=False):
```
Comment (wangch079) [inline connectors/sources/sharepoint_online.py:204]: Why is `debug` used here?

File: connectors/sources/sharepoint_online.py [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+    name = "Sharepoint Online"
+    service_type = "sharepoint_online"
+
+    def __init__(self, configuration):
+        super().__init__(configuration=configuration)
+
+        self._client = None
+
+    @property
+    def client(self):
```
Comment (wangch079) [inline connectors/sources/sharepoint_online.py:482]: You can use [@cache](https://docs.python.org/3/library/functools.html#functools.cache)

see example: 
https://github.com/elastic/connectors-python/blob/7539174dfb0a35a1129e0b992e3c5702080b7e98/connectors/source.py#L588

File: connectors/sources/sharepoint_online.py [hunk 4]
```diff
@@ -0,0 +1,740 @@
...
+        # Check that we can log in into Graph API
+        await self.client.graph_api_token.get()
+
+        # Check that we can log in into Sharepoint REST API
+        await self.client.rest_api_token.get()
+
+
+        # TODO: check that tenant_name is valid
+
+        # Check that we at least have permissions to fetch sites and actual site names are correct
```
Comment (wangch079) [inline connectors/sources/sharepoint_online.py:543]: It seems this is out of the scope of `validate_config`.

File: connectors/sources/sharepoint_online.py (hunk 2 above)
Comment (artem-shelkovnikov) [reply connectors/sources/sharepoint_online.py:204]: Cause it's a WIP and I sometimes push things that won't go into production :D

File: connectors/sources/sharepoint_online.py (hunk 3 above)
Comment (artem-shelkovnikov) [reply connectors/sources/sharepoint_online.py:482]: Thanks, will do!

File: connectors/sources/sharepoint_online.py [hunk 5]
```diff
@@ -0,0 +1,811 @@
...
+            "_attachment": await self._download_content(
+                partial(
+                    self.client.download_drive_item,
+                    drive_item["parentReference"]["driveId"],
+                    drive_item["id"],
+                )
+            ),
+        }
+
+    async def _download_content(self, download_func):
```
Comment (navarone-feekery) [inline connectors/sources/sharepoint_online.py:783]: I love that this is abstracted from `get_content` and `get_attachment` now 🔥

**Review by navarone-feekery — APPROVED** [review]
LGTM, there are some nitpicky grammatical points I would like to fix in the readme, but in the interest of having this merged, I will make my own PR after instead.
Nice work!!

**Pull request merged**